crypto/init.c

```diff
@@ -535,18 +496,8 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
             return 1;
     }
 
-    /*
-     * Now we don't always set up exit handlers, the INIT_BASE_ONLY calls
-     * should not have the side-effect of setting up exit handlers, and
-     * therefore, this code block is below the INIT_BASE_ONLY-conditioned early
-     * return above.
-     */
-    if ((opts & OPENSSL_INIT_NO_ATEXIT) != 0) {
-        if (!RUN_ONCE_ALT(&register_atexit, ossl_init_no_register_atexit,
-                          ossl_init_register_atexit))
-            return 0;
-    } else if (!RUN_ONCE(&register_atexit, ossl_init_register_atexit)) {
-        return 0;
+    if (opts & OPENSSL_INIT_NO_ATEXIT) {
+        no_cleanup = 1;
```

Comment:
Do we actually need or want this? The OPENSSL_INIT_NO_ATEXIT was needed because using the atexit caused crashes, etc. The use of library destructors should not have such consequences because once the library is unloaded it cannot be used anymore anyway. And we have the pinshared option to keep the library loaded until the process is terminated.

Reply:
I kept this option in to keep tests happy. There is a test for this:
https://github.com/openssl/openssl/pull/24148#discussion_r1567630865
the test currently fails on windows. I'm still investigating failure. I'm fine with removing the option, but the option is also documented in `OPENSSL_INIT_CRYPTO()`. I was unsure about keeping it/dropping it because I don't know how deprecation/removal process of documented API works here.

Reply:
We need to keep the OPENSSL_INIT_NO_ATEXIT flag but it can be a no-op. Either that or you would have to have the same no_cleanup handling in destructors on all the possible platforms where you have them.